### DISCUSSION TITLE
CI/CD 101, where am I going wrong?

### DISCUSSION THREAD

**Discussion by woter1832** [community/community#26859] [Actions]
<p>Hi,</p>
<h1>Creating releases / atefacts / outputs / publishing from CI.</h1>
<p>I may be going around this completely the wrong way, hence why I cannot seem to find a solution.</p>
<p>I have Nexus artefact repository set up to take Docker images.</p>
<p>In my pipeline, I build the container in a non-master branch. When I am happy with it, I want to create a PR, which once approved will push to master branch and publish the container image to Nexus.</p>
<p>I am having a hard time getting GitHub workflows to do the build and publish in different “phases”.</p>
<h3>My ideas:</h3>
<p>I’m aware of GitHub releases, but AFAIK, they don’t support container images. I would need to use a container repo such as <a href="http://docker.io" rel="noopener nofollow ugc">docker.io</a> or my own (Nexus).</p>
<p>Half the problem seems to be the <code>on:</code> statement. It happens at the start of the workflow. Is there a way to have multiple workflows with their own <code>on:</code> settings? This way the first one can build from non-master branches and the second one can build from master.</p>
<p>To achieve this, one way I am trying is to split my workflow into two jobs: <code>build</code> and <code>publish</code>. I’ve set <code>if: github.event.pull_request.merged</code> on the <code>publish</code> job but it never triggers. My <code>on:</code> statement looks like:</p>
<pre><code class="lang-auto">on:
  push:
    branches: [ develop, master ]
  pull_request:
    branches: [ master ]
</code></pre>
<p>Even echoing out <code>github.event.pull_request.merged</code>, never returns anything.</p>
<p>I’m sure this is CI/CD 101, but I can’t get it to work. Any help would be greatly appreciated.</p>
<p>T.I.A</p>

**Comment by lee-dohm** [comment, marked as answer]
<aside class="quote no-group" data-username="woter1832" data-post="1" data-topic="133567">
<div class="title">
<div class="quote-controls"></div>
 woter1832:</div>
<blockquote>
<p>Is there a way to have multiple workflows with their own <code>on:</code> settings?</p>
</blockquote>
</aside>
<p>Yes, you can have as many different workflows as you like and each one can have its own <code>on:</code> settings. What it looks like you’re trying to do is have it all in one file. If you split the two concepts into two separate files (each one containing a single workflow), then things should work better for you.</p>
<p>For example, in my <code>select-matching-issues</code> Action I have <a href="https://github.com/lee-dohm/select-matching-issues/tree/490b7362747e5862d4d9031c30411231ebc5d798/.github/workflows">two separate workflows</a> for testing and publishing.</p>
<p>Is that something that would help?</p>

**Comment by woter1832** [comment]
<p>Thank you very much <a class="mention" href="/lee-dohm">@lee-dohm</a>.<br>
Obvious when you know how.</p>
<p>This free training course also helped: <a href="https://lab.github.com/githubtraining/create-a-release-based-workflow" rel="noopener nofollow ugc">https://lab.github.com/githubtraining/create-a-release-based-workflow</a></p>
